Use mbedTLS hostname verification for LibGit2

omus opened the pull request:
Corrects issues with hostname verification failing when the SSL subjectAltName is not specified and the CN should be used. Specifically the original `verify_server_cert` [check of the subjectAltName](https://github.com/JuliaLang/julia/compare/cv/libgit2-ssl-verify?expand=1#diff-7b94fb465bb67b40e70933b8f97f7d65L523) would fail since `cert->subject_alt_names` would return a single empty string. Additionally the [CN check](https://github.com/JuliaLang/julia/compare/cv/libgit2-ssl-verify?expand=1#diff-7b94fb465bb67b40e70933b8f97f7d65L548) would perform a string comparison against the entire subject line which would cause to always fail.

The fix was to use the built in mbedTLS verification which should be much more reliable.

I ended up correcting the original patch instead of creating a new patch. Let me know if a new patch is preferred.

omus committed:
Use mbedTLS hostname verification for LibGit2

Corrects issues with hostname verification failing when the SSL subjectAltName is not
specified and the CN should be used.

omus commented:
Also an issue on Julia 0.5. I'll verify that the patch can be backported or make a new PR specifically for 0.5.

omus commented:
cc: @wildart

tkelman commented:
Downside of modifying the existing patch in-place is that people won't see the update unless they do a clean build or distclean-libgit2. How big of an issue is this, and can we test it in any way?

omus commented:
Mostly this is just a problem for those using private Git servers with self-signed certificates or being their own CA. It seems uncommon to not include the subjectAltName.

I should be able to make some tests for this using `openssl s_server` or something similar.

omus committed:
Add test for mbedTLS verification

omus commented:
Added a test to make sure that certificates with only a common name are verified correctly. I ended up using `openssl s_server` as the SSL server in the test since we already require it to make the certificate. Unfortunately the basic `openssl s_server` cannot actually host a Git repository which means even with a successful test we get an exception. Since the exception is different however we can tell the difference between a SSL exceptions and "repo cloning" exceptions.

omus commented:
@tkelman currently the tests are failing in Travis. I believe this is due to libgit2 not being re-compiled. Is there something specific I need to do for Travis or is this the problem you mentioned with updating the patch in-place?

omus committed:
Made mbedTLS verification a seperate patch

tkelman commented on test/libgit2.jl:
```diff
@@ -915,4 +915,74 @@ mktempdir() do dir
+
+                try
+                    # The generated certificate is normally invalid
+                    run(`$(Base.julia_cmd()) -e $code`)
```
startup-file=no just in case on these

omus committed:
Disable startup-file

omus commented:
Travis CI error seems unrelated as the libgit2 tests are passing. Feel free to squash and merge.

tkelman commented on test/libgit2.jl:
```diff
@@ -915,4 +915,75 @@ mktempdir() do dir
+        if is_linux()
+            try
+                # OpenSSL needs to be on the path
+                openssl_command = strip(readstring(`which openssl`))
```
why not something trivial like `openssl version` instead of relying on `which` ?

omus replied:
Good idea

tkelman commented:
I'll want to run this through the buildbots to make sure it passes on non-ubuntu distros

omus committed:
Remove use of which

tkelman commented on test/libgit2.jl:
```diff
@@ -915,4 +915,76 @@ mktempdir() do dir
+                errfile = joinpath(root, "error")
+                reponame = "https://$hostname:4433/Example.jl"
+                code = """
+                    dest = tempname()
```
this should be inside the mktempdir folder

tkelman commented on test/libgit2.jl:
```diff
@@ -915,4 +915,76 @@ mktempdir() do dir
+
+                    rm(errfile)
+
+                    # Specify that Julia use onl the custom certificate. Note: we need to
```
use only

omus committed:
Changes from PR review

tkelman commented:
As I thought it might, fails tests on the centos 7 buildbot: https://build.julialang.org/builders/build_centos7.1-x64/builds/2865/steps/shell_2/logs/stdio

omus commented:
Good find. I should be able to fall back to using `localhost` as the common name if the hostname isn't mapped to the loopback device.

omus committed:
Use localhost when hostname isn't set to loopback

tkelman commented:
that also failed:
https://build.julialang.org/builders/build_centos7.1-x64/builds/2869/steps/shell_2/logs/stdio

omus commented:
I didn't expect that `getaddrinfo` would throw an exception when it cannot lookup a hostname. Easy to fix.

omus committed:
Handle getaddrinfo exception

tkelman commented on test/libgit2.jl:
```diff
@@ -928,15 +928,24 @@ mktempdir() do dir
+            end
+
+            if isempty(common_name)
                 warn("Unable to determine hostname that maps to the loopback address")
```
should this also say "Skipping hostname verification tests" or something to that effect?

omus replied:
Doesn't hurt to add more info

omus committed:
[ci skip] Improve warning message

tkelman commented on deps/libgit2.mk:
```diff
@@ -84,6 +84,11 @@ $(LIBGIT2_SRC_PATH)/libgit2-mbedtls-writer-fix.patch-applied: $(LIBGIT2_SRC_PATH
 		patch -p1 -f < $(SRCDIR)/patches/libgit2-mbedtls-writer-fix.patch
 	echo 1 > $@
 
+$(LIBGIT2_SRC_PATH)/libgit2-mbedtls-verify.patch-applied: $(LIBGIT2_SRC_PATH)/source-extracted
```
this should depend on any previous patches that touch the same files, so they happen in order even with parallel make

omus replied:
How do I accomplish that? Do I just add the following?
`.../source-extracted | $(LIBGIT2_SRC_PATH)/libgit2-mbedtls.patch-applied`

tkelman replied:
yep that should do it

omus replied:
Ok. The patch `libgit2-mbedtls-writer-fix.patch` also modifies the same file as the new patch but a separate section of the file. Since the two patches do not conflict the order which they are applied shouldn't matter so I'll just do the dependency on `libgit2-mbedtls.patch`

tkelman replied:
it would be better to make the dependencies explicit, there would be a race condition if both patches get applied at the same time, one of the applications could get overwritten by the other

omus replied:
Ok, I'll make the dependency explicit.

tkelman commented:
passing on (most of) the buildbots now, so address the two most recent comments then we can merge this

omus committed:
Add libgit2-mbedtls-hostname.patch dependency

libgit2-mbedtls-hostname patch directly depends on the libgit2-mbedtls
patch. The libgit2-mbedtls-writer-fix patch also modifies the
mbedtls_stream.c file seperate section of the file. To ensure that the
patches are always applied in the same order we'll make the hostname
patch dependent on the writer-fix patch.

tkelman reviewed (approved):
lgtm if CI is happy. should probably be squashed (with the [ci skip] edited out)

omus commented:
Definitely should be squashed.
